---
title: "DESeq2 Tutorial"
author: "Charlie George & Modified by Devika Agarwal"
date: "`r Sys.Date()`"
output: html_document
---

Useful resources 

- **DESeq2 Vignette** - http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
- **Bioconductor RNAseq Workflow Book** - http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html


# install packages 

- If i didn't already have these installed in my environment - I would delete the # on lines 19-23 and 26-39 and run the code 

```{r}
# from cran 
# install.packages('tidyverse')
# install.packages("RColorBrewer")
# install.packages('pheatmap')
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# from bioconductor
# BiocManager::install("ashr")
# BiocManager::install("apeglm")
# BiocManager::install("DESeq2")
# BiocManager::install("org.Mm.eg.db")
```

# load packages

- what do these packages do? Add comments to remind you what they do 
```{r}
library("pheatmap")   
library("RColorBrewer")
library(tidyverse)     
library(DESeq2)      
library(ashr)          
library(apeglm)      
library(org.Mm.eg.db)
```

# 1 Prepping our input data to make deseq2dataset object

## Import counts table

- We made this (or a very similar version) using featurecounts in week1 on the cluster
- raw data -> fastqc -> hisat2 -> featurecounts 

```{r}
# Import counts data to variable called featurecounts_table
# The file has a comment line we need to ignore at top of file 
featurecounts_table <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/allsamples_counts.tsv.gz", comment = "#")


# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)


# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
counts_table <- featurecounts_table %>%
                  column_to_rownames("Geneid") %>%
                  dplyr::select(starts_with("ERR") & ends_with(".bam")) %>%
                  as.matrix()

counts_table

# check the dimensions of this counts_table table using dim() -> should have 39,179 genes and 8 cols 


```


# Right now lets import and set up our table of sample information

```{r}
sample_info <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/PRJEB18572_sample_info.tsv")
sample_info

# What class is the cell_type currently? 

class(sample_info$cell_type)


# We want to change it to a factor with 2 levels (the hypothesis we will test)
# Which should be our reference level? 

sample_info$cell_type <- factor(sample_info$cell_type, levels = c("CD4", "CD8"))


# what other columns might we want to include in our design that we want to change to factors? 

sample_info$genotype <- as.factor(sample_info$genotype)
sample_info$replicate <- as.factor(sample_info$replicate)
sample_info$day_collected <- as.factor(sample_info$day_collected)




```

# Check the sample names are the same and in the same order in the sample_info and counts_table 

```{r}
# Column names of the count table and row names of the sample info table must be in the same order

identical(colnames(counts_table), rownames(sample_info))

colnames(counts_table) == rownames(sample_info)

colnames(counts_table)
rownames(sample_info)


# Check this 

# make sample_info row names from sample_id 

rownames(sample_info) <- sample_info$sample_id
rownames(sample_info)

# remove .bam from colnames 
colnames(counts_table)

colnames(counts_table) <- counts_table %>%
    colnames() %>%
    str_replace(".bam","")

colnames(counts_table) <- str_remove(colnames(counts_table) , ".bam")


colnames(counts_table) == rownames(sample_info)
```

# Generate DESeqDataSet object

```{r}
# look up function to use from slides, find help and create an object called dds
dds <- DESeqDataSetFromMatrix(countData = counts_table,
                              colData = sample_info,
                              design = ~ 1)

rowData(dds) <- gene_info

  
```

## Investigate the dds object 

```{r}
# what does it look like? - e.g. use str(dds)
str(dds)

# how to access counts
assays(dds)

assay(dds, "counts")

# access colData i.e the sample info table
colData(dds)

# access rowData
rowData(dds)

# access the design
design(dds)

# does it have any rowRanges()?
rowRanges(dds)

# what are its rownames and colnames?
rownames(dds)
colnames(dds)

# what attributes does it have?
attributes(dds)

```

## How to subset a DESeqDataSet Object 

```{r}
# practice subsetting dds object

# select first 3 columns
dds_subset3 <- dds[,3]

```

# Filter out unexpressed genes 

- It's good practice to filter out genes with 0 counts across all samples at the start 
- This reduces the size of the object and makes computations quicker without affecting the results as these genes are filtered out anyway in the analysis 
- Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.


```{r}
# filter out genes that have 0 counts across all samples and save as an object called `dds_filtered`
rowsums_outcome <- rowSums(assay(dds, "counts")) > 0

dds_filtered <- dds[rowsums_outcome,]
remove(rowsums_outcome)


dds_filtered2 <- dds[rowSums(assay(dds, "counts")) > 0,]

assay(dds_filtered, "counts") == assay(dds_filtered2, "counts")

# check dims of dds_filtered  
dim(dds)
dim(dds_filtered)

# can also be more stringent in filtering depending on the dataset, filter out low expressing genes with less than 10 counts across 3 samples or higher for example
# at least 3 samples with a count of 10 or higher
#keep <- rowSums(counts(dds) >= 10) >= 3
#dds <- dds[keep,]
#nrow(dds)  
  
```

# Perform exploratory analysis on dds_filtered object 

1. transform dds_filtered object using `rlog()` functions
    - name your new variables `rlog_transformed_dds` 
    - we want to set `blind = TRUE` as because we are doing a QC step we do not want the function to take into account the experimental design, we want it to be an unbiased look at how the samples cluster
    - check the class of the rlog_transformed_dds objects 
    - you can look at the tranformed counts table using `assay()` 

NOTE we only use this tranformed data for QC not for downstream analysis or differential expression itself!!


```{r}
#transform data using `rlog()` functions

rlog_transformed_dds <- rlog(dds_filtered, blind = TRUE)

#to add in to the same object
assay(dds_filtered, "rlog transformed") <- rlog(dds_filtered, blind = TRUE)


# Notes on `blind` parameter: 
  # blind = TRUE is default
  # blind = TRUE is not the appropriate choice if you expect many or majority of genes to have large differences in counts explainable by   experimental design
  # Mike Love recommends blind = FALSE for visualisation, clustering and machine learning but blind = TRUE for QC
  # Experimental design is only used for calculating the gene-wise dispersion estimates in order to fit a trend line through the dispersions over the mean

  
#check the class of these variable objects 
class(rlog_transformed_dds)
  
#get and look at the tranformed counts table using `assay()`
assay(rlog_transformed_dds)

```

# Optional: If you have time do the tranformation again but using the vst() function 

- As a guide 
    - rlog - more robust than VST when size factors vary widely (10x)
    - VST - less sensitive to high count outliers than rlog, quicker - recommended most often 

```{r}
# get vst transformed data using vst() 
vst_transformed_data <- vst(dds_filtered, blind = TRUE)

```


# Now plot heatmap of sample-sample correlations using rlog transformed count data and cor()


```{r}
# get correlation matrix of the rlog_transformed counts matrix to get correlations between samples 

library(grDevices)
assays(rlog_transformed_dds)
names(assays(rlog_transformed_dds))

cor_matrix <- cor(assay(rlog_transformed_dds), method = )

# use pheatmap to plot the correlation matrix 

pheatmap(cor_matrix, color = colorRampPalette(c("navy", "white", "firebrick3"))(50))

```

# try ploting a heatmap of sample distances 

```{r}
# get sample distances -> Annotated what this code is doing?

#the dist function computes and returns the distance matrix computed by using the specified distance measure to compute the distances between the rows of a data matrix. Therefore we need to use the t to TRANSFORM the matrix so that it looks at the distances between the samples based on the genes (rather than the distances between the genes based on the samples)

sampleDists <- dist(t(assay(rlog_transformed_dds))) 
sample_distances_matrix <- as.matrix(sampleDists) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)
```

What do you notice here!? what does it tell us about our data? 

Answer: There are two duplicates (-82 and -88) and (-83 and -89). To check this, run identical or unique functions on the counts matrices of these samples



## You can subset a dds object and interact with it like you would a DataFrame

```{r}
# remove the two suspect columns using [,] notation - remove suspicious samples 
rlog_transformed_dds_subset <- rlog_transformed_dds[,1:6]
#alternative method: rlog_transformed_dds_subset <- rlog_transformed_dds[, -(7:8)]

# lets replot the correlation heatmap with the subsetted data - copy code and adjust from above

sampleDists <- dist(t(assay(rlog_transformed_dds_subset))) 
sample_distances_matrix <- as.matrix(sampleDists) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)

colData(rlog_transformed_dds_subset)
```

What do you notice now about the data and the clustering? There are two different groups (two different cell types)


# Now lets use the plotPCA function from DESeq2 to plot PCA of rlog_transformed_data_subset

```{r}
# plot pca of rlog_transformed_dds using plotPCA()

colData(rlog_transformed_dds_subset)

plotPCA(rlog_transformed_dds_subset, intgroup = "replicate") + coord_cartesian()

# what does the intgroup parameter do? have a go at changing it...


# how many variable genes is it using for pca? 


# plot the pca for all genes 

```




What does it tell us about our sample?

Answer: 


------------- This is the end of our Exploratory data analysis -----------------

We now go back to our original dds object with the raw counts and can update it with anything we need to from the exploratory analysis 
    e.g. - update design with variables in sample info table we want to regress out 
         - remove samples (poor quaility or those with name switches etc)
         
# Revisiting our inital data 

What do we need to do to our data??? 

## Update dds_filtered object if needed - call it dds_filtered_updated

```{r}

# do we need to remove/edit samples? 
dds_filtered_updated <- dds_filtered[,1:6]


# do we want to update the design? Is this nessecary in this case?
# e.g. design(dds) <- ~replicate + condition

colData(dds_filtered_updated)

design(dds_filtered_updated) <- ~ cell_type

```


# Now we can start our actual DE!!!! yay 

Run deseq2 modelling using one function (DESeq())

```{r}

# look at dds slots then run deseq2 function 
# what has happened to your dds slots?

# run deseq2 modelling on the updated dds object in one function using DEseq() 
dds_filtered_updated <- DESeq(dds_filtered_updated)

# then extract the results
results_obj <- results(dds_filtered_updated)

# sumarise results 
summary(results_obj)

```

Have a look at the outputs 

# Now we will run the analysis step by step to see whats going on 

## 1. Estimate size factors using estimateSizeFactors()

```{r}
# Calculate the size factors for each sample using estimateSizeFactors()
dds_filtered_updated <- estimateSizeFactors(dds_filtered_updated)

# Access the size factors from the dds_filtered_updated object
sizeFactors(dds_filtered_updated)

# Do you see any systematic differences in size factors between groups?

#CD4 group overall has higher size factors than CD8 group (so relatively less counts than CD8 group)


```

# Optional: Plot size factors as barchart using ggplot
```{r}

size_factors <- sizeFactors(dds_filtered_updated)
size_factors <- as.data.frame(size_factors)
size_factors$sample_id <- rownames(size_factors)

colData_df <- as.data.frame(colData(dds_filtered_updated))
ggplot(colData_df, aes(x =sample_id, y = sizeFactor, fill = cell_type)) + geom_col()
```

## Get counts table normalised using size factors

```{r}
# get counts table normalised using size factors - call it normalized_counts
normalized_counts <- counts(dds_filtered_updated, normalized = TRUE)

# check the class of the output
class(normalized_counts)
head(normalized_counts)
```

# 3. Estimate dispersions for each gene

```{r}
# Obtain dispersion estimates for each gene using estimateDispersions() 

dds_filtered_updated <- estimateDispersions(dds_filtered_updated)

# Plot the per-gene dispersion estimates (DESeq2 has a helper function for this)
plotDispEsts(dds_filtered_updated)

```


## 4. Perform Wald test

```{r}
# Tests for significance of coefficients in a negative binomial GLM, using previously calculated size factors and dispersion estimates and WaldTest

dds1 <- nbinomWaldTest(dds_filtered_updated)
results(dds1)
summary(dds1)
mcols(dds1)

```


# So we have the actual steps that we've gone through to get to DE in one place generate the dds object again and use the DESeq() function to perform steps 1-3 in one go

```{r}
# generate dds object 
dds <- DESeqDataSetFromMatrix(countData = counts_table,
                              colData = sample_info,
                              rowData = gene_info,
                              design = ~ 1)


# filter rows with 0 counts 
rowsums_outcome <- rowSums(assay(dds, "counts")) > 0

dds <- dds[rowsums_outcome,]
remove(rowsums_outcome)

# remove unwanted samples 

dds <- dds[,1:6]

#add in right design
design(dds) <- ~ cell_type

# run DESeq() 
dds <- DESeq(dds)

ddsresults <- results(dds)

```

# Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify? 

ddsresults
summary(ddsresults)
class(ddsresults)

summary(dds)
class(dds)

```

# Optional - if we have time: Access the coefficients of the Negative Binomnial GLM – see vingette("DESeq2")
```{r}

```

# Plot normalised counts for ENSMUSG00000041779
 
 checkout the 'plotCounts()` function
 
```{r}
# Plot normalised counts for gene ENSMUSG00000041779
plotCounts(dds, "ENSMUSG00000041779", intgroup = "cell_type")


```

# Access the results table for the comparison between CD8 and CD4 T cells . Youu can also try to CD4 vs CD8 (i.e. switch the reference) - you could also try this for replicate 1 vs replicate 3 as a control 

checkout `resultsNames()`  & `results()`

```{r}
resultsNames(dds)

ddsresults <- results(dds, name = "cell_type_CD8_vs_CD4")
ddsresults
```

Table should contain log2 fold changes and p-values for all genes

# Plot a histogram of the raw and BH-adjusted p-values using ggplot – do they look as expected? for CD8 vs CD4

```{r}
ggplot(as.data.frame(ddsresults), aes(x = pvalue, y = padj, na.rm = TRUE)) + geom_point() + geom_abline(slope = 1, intercept = 0, color = "red")

ggplot(as.data.frame(ddsresults)) + geom_histogram(aes(x = pvalue), fill = "blue", alpha = 0.5) + geom_histogram(aes(x = padj), fill = "red", alpha = 0.5)

```


# Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}
plotMA(ddsresults, alpha = 0.05) + abline(h=c(-1,1), col="red", lwd=2)

```

# Shrink the log2 FC values using the normal, apeglm and ashr methods -> see the vignette – which shall we choose? 

```{r}

# call your results table  called shrunkenLFC_dds_results_obj

shrunkenLFC_dds_results_obj <- lfcShrink(dds, coef="cell_type_CD8_vs_CD4", type="apeglm")


plotMA(shrunkenLFC_dds_results_obj, alpha = 0.05) + abline(h=c(-1,1), col="red", lwd=2)

cowplot::
  
```


# Generate a results table (one shrinkage method) containing gene symbols and (if time) entrez ids 

```{r}

# convert shrunkenLFC_dds_results_obj to a dataframe and call it shrunkenLFC_dds_results_df

shrunkenLFC_dds_results_df <- as.data.frame(shrunkenLFC_dds_results_obj)

# Use the org.Mm.eg.db package to get gene symbols for the ensembl gene ids (hint:you did this on bioconductor day) 
# use mapIds() to get 1:1 mapping for SYMBOL, ENTREZ and MGI

library(org.Mm.eg.db)

columns(org.Mm.eg.db)
keytypes(org.Mm.eg.db)

shrunkenLFC_dds_results_df$Ensembl <- rownames(shrunkenLFC_dds_results_df)

shrunkenLFC_dds_results_df$Gene_symbol <- mapIds(org.Mm.eg.db, 
                                                 keys = shrunkenLFC_dds_results_df$Ensembl, 
                                                 column = "SYMBOL", 
                                                 keytype = "ENSEMBL")

shrunkenLFC_dds_results_df$ENTREZID <- mapIds(org.Mm.eg.db, 
                                              keys = shrunkenLFC_dds_results_df$Ensembl, 
                                              column = "ENTREZID", 
                                              keytype = "ENSEMBL")

shrunkenLFC_dds_results_df$MGI <- mapIds(org.Mm.eg.db, 
                                         keys = shrunkenLFC_dds_results_df$Ensembl, 
                                         column = "MGI", 
                                         keytype = "ENSEMBL")

# How many Ensembl IDs are not assigned an gene symbol and how many gene symbols are duplicated?

sum(is.na(shrunkenLFC_dds_results_df$Gene_symbol)) #answer = 4197
duplicated_symbols <- shrunkenLFC_dds_results_df$Gene_symbol %>% duplicated() %>% [] #




keytypes(org.Mm.eg.db)      #to remind us of what keytypes are in this particular db

# use mapIds() to get 1:1 mapping for SYMBOL, ENTREZ and MGI

# add mapId vectors to  shrunkenLFC_dds_results_df
```


# Export all the results to CSV

```{r}
write_csv(shrunkenLFC_dds_results_df, "CD8_vs_CD4_DE_results_lfcshrinkage_allgenes.csv")
```


# Filter res_df and save as CSV

```{r}
#Filter the results table for padj < 0.05 and absolute log2 FC > 1, and write to a CSV file
res_0.05 <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > 1)

#Get upregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC > 1, and write to a CSV file
res_0.05_up <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > 1)

#Get doownregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC < 1, and write to a CSV file
res_0.05_down <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -1)

write_csv(res_0.05, 
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1.csv")

write_csv(res_0.05_up,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_up.csv")

write_csv(res_0.05_down,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_down.csv")
```


Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account. 

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....

